kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/json/JsonRecordWriterProvider.java

```diff
@@ -51,14 +52,15 @@
 
   @Override
   public String getExtension() {
-    return EXTENSION;
+    return EXTENSION + storage.conf().getCompressionType().extension;
   }
 
   @Override
   public RecordWriter getRecordWriter(final S3SinkConnectorConfig conf, final String filename) {
     try {
       return new RecordWriter() {
         final S3OutputStream s3out = storage.create(filename, true);
+        final OutputStream out = s3out.wrappedForCompression();
         final JsonGenerator writer = mapper.getFactory()
                                          .createGenerator(s3out)
```

Comment:
I think this needs to be changed to `.createGenerator(out)`. If the SinkRecord value is already encoded as JSON (like when this sink is configured with `"value.converter": "org.apache.kafka.connect.json.JsonConverter"`), it is written to S3 uncompressed due to the conditional in `write()`. See line 82. The output file in S3 has gzip headers, but an uncompressed body.

Reply:
Nice catch. I think you're right. Will push a change and make sure the tests are exercising this.